controller/src/main/java/org/jboss/as/controller/AttributeMarshaller.java

```diff
@@ -105,6 +105,33 @@ protected String asString(ModelNode value) {
         }
     }
 
+    private static class ObjectMarshaller extends DefaultAttributeMarshaller {
+        private final boolean marshallSimpleTypeAsAttribute;
+
+        private ObjectMarshaller(boolean marshallSimpleTypeAsAttribute) {
+            this.marshallSimpleTypeAsAttribute = marshallSimpleTypeAsAttribute;
+        }
+
+        @Override
+        public void marshallAsElement(AttributeDefinition attribute, ModelNode resourceModel, boolean marshallDefault, XMLStreamWriter writer) throws XMLStreamException {
+            if (resourceModel.hasDefined(attribute.getName()) && attribute instanceof ObjectTypeAttributeDefinition) {
+                AttributeDefinition[] valueTypes = ((ObjectTypeAttributeDefinition) attribute).getValueTypes();
+                writer.writeStartElement(attribute.getXmlName());
+                for (AttributeDefinition valueType : valueTypes) {
+                    if(resourceModel.get(attribute.getName()).hasDefined(valueType.getName())) {
+                        ModelNode handler = resourceModel.get(attribute.getName());
+                        if(!marshallSimpleTypeAsAttribute) {
```

Comment:
I find this variable name confusing given it results in a call to "marshallAsElement". It needs comments, and the Javadoc for ELEMENT_ONLY_OBJECT and ATTRIBUTE_OBJECT needs to be a lot more explicit about what results will happen if those marshallers are used.